src/bodse/scheduler.py

```diff
@@ -0,0 +1,704 @@
+# -*- coding: utf-8 -*-
+"""
+Functionality for automating the downloading and processing of BODS data.
+
+The scheduler will perform the following tasks on a regular schedule:
+- Download the GTFS schedule from BODS
+- Download AVL data from BODS for a set period
+- Calculate an adjusted GTFS schedule using AVL data
+- Upload all datasets to a PostgreSQL database
+"""
+
+##### IMPORTS #####
+
+# Built-Ins
+import calendar
+import datetime
+import enum
+import logging
+import pathlib
+import re
+import time
+import traceback
+from collections import abc
+from typing import Any, Optional
+
+# Third Party
+import pydantic
+from caf.toolkit import config_base, log_helpers
+from pydantic import dataclasses, types
+
+# Local Imports
+import bodse
+from bodse import database, request, teams, timetable
+from bodse.avl import adjust, avl
+from bodse.avl import database as avl_db
+
+##### CONSTANTS #####
+
+LOG = logging.getLogger(__name__)
+WAIT_TIME = 3600 * 12  # 0.5 days
+
+
+##### CLASSES & FUNCTIONS #####
+
+
+class Day(enum.IntEnum):
+    """Days of the week."""
+
+    MONDAY = calendar.MONDAY
+    TUESDAY = calendar.TUESDAY
+    WEDNESDAY = calendar.WEDNESDAY
+    THURSDAY = calendar.THURSDAY
+    FRIDAY = calendar.FRIDAY
+    SATURDAY = calendar.SATURDAY
+    SUNDAY = calendar.SUNDAY
+
+    @property
+    def name_and_number(self) -> str:
+        """Return name and number of day."""
+        return f"{self.name} ({self.value})"
+
+    @classmethod
+    def _missing_(cls, value: Any) -> "Day":
+        try:
+            value = int(value)
+        except ValueError:
+            pass
+
+        if isinstance(value, int):
+            for i in cls:
+                if value == i.value:
+                    return i
+
+        else:
+            value = str(value)
+            upper = value.upper().strip()
+
+            if upper in [i.name for i in cls]:
+                return cls[upper]
+
+            for i in cls:
+                if i.name[:3] == upper:
+                    return i
+
+        raise ValueError(
+            f"{value!r} is not a valid {cls.__name__} should be one of "
+            + ", ".join(i.name_and_number for i in cls)
+        )
+
+
+@dataclasses.dataclass
+class TaskParameters:
+    """Parameters for scheduled BODSE tasks."""
+
+    gtfs_folder: pydantic.DirectoryPath
+    avl_download_time: avl.DownloadTime
+    max_timetable_age_days: int = pydantic.Field(7, ge=0)
+    max_avl_age_days: int = pydantic.Field(7, ge=0)
+    avl_database_delete_age_days: int = pydantic.Field(30, ge=0)
+
+
+class SchedulerConfig(config_base.BaseConfig):
+    """Parameters for running BODSE scheduler."""
+
+    output_folder: types.DirectoryPath
+    task_parameters: TaskParameters
+    database_parameters: database.DatabaseConfig
+    api_auth_config: request.APIAuth
+    teams_webhook_url: Optional[pydantic.HttpUrl] = None
+
+    run_months: int = pydantic.Field(2, ge=1, le=12)
+    run_month_day: int = pydantic.Field(5, ge=1, le=28)
```

Comment:
Another question!

Assuming max day of 28 here is due to Feb. 

If there was to be support for running on later date in different month I assume this would require a new validator to check that the `run_month_day` would be a valid value given the month used? 

I don't think this is an issue though.

Reply:
Yeah so that it could run in Feb, if the days was above the number in the current month then it wouldn't be able to run within that month and therefore might need to wait an additional month